### ISSUE TITLE
Automation Condition can't handle assets with optional outputs correctly 

### ISSUE THREAD

**Issue by Zan-L** [dagster-io/dagster#24389]
### Dagster version

1.8.6

### What's the issue?

Suppose we have 3 assets, (Root -) A - B - C, where the output of B is optional, and let's say we have eager policies on all. When the condition is met, a run of 3 assets will be triggered. 

When B does not yield output, C will be skipped in that run. However, for the next next (yes double next) sensor tick, another run of (B, C) will be spawned, which should not happen. 

The root cause lies in the condition, any_dep_updated().since(newly_requested() | newly_updated()), which evaluates to false in the next sensor tick due to newly_request(), but changes to true in the next tick after that, because for B, dep A is updated but B itself is not updated as a result of its lack of output. 

The fix would be to define a new condition that correctly evaluates to true after a run. I tried any_dep_updated().since((~in_progress()).newly_true()), but due to how (~in_progress()).newly_true() is evaluated, it only works sometimes. Maybe a new event condition is worth creating. 

### Deployment type

None

### Message from the maintainers

Impacted by this issue? Give it a 👍! We factor engagement into prioritization.

**Comment by OwenKephart** [comment]
Hi @Zan-L, thanks for the detailed report here. The ideal would be if we had an explicit "asset skipped" event that we could listen for, which would allow us to detect "newly_skipped()" or something of the sort. However, in the absence of such an event, and building on the idea you had, here's an idea that does work:

```python
import time

import dagster as dg


def get_automation_condition() -> dg.AutomationCondition:
    return (
        dg.AutomationCondition.in_latest_time_window()
        & (
            dg.AutomationCondition.newly_missing() | dg.AutomationCondition.any_deps_updated()
        ).since(
            dg.AutomationCondition.newly_requested()
            | dg.AutomationCondition.newly_updated()
            | dg.AutomationCondition.in_progress()
        )
        & ~dg.AutomationCondition.any_deps_missing()
        & ~dg.AutomationCondition.any_deps_in_progress()
        & ~dg.AutomationCondition.in_progress()
    ).with_label("fancy_eager")


@dg.asset
def root() -> None: ...


@dg.asset(
    deps=[root],
    automation_condition=get_automation_condition(),
)
def a() -> None:
    time.sleep(15)


@dg.asset(
    deps=[a],
    output_required=False,
    automation_condition=get_automation_condition(),
)
def b():
    time.sleep(15)
    if False:
        yield dg.Output(None)


@dg.asset(
    deps=[b],
    automation_condition=get_automation_condition(),
)
def c() -> None: ...


defs = dg.Definitions(
    assets=[root, a, b, c],
    sensors=[
        dg.AutomationConditionSensorDefinition(
            "ac_sens",
            asset_selection="*",
            default_status=dg.DefaultSensorStatus.RUNNING,
            minimum_interval_seconds=5,
        )
    ],
)
```

The subtle bit here is that it updates the logic that checks for parent updates to ignore any updates that happen while the asset is in progress (basically swapping out `since_last_handled` for `since(handled | in_progress)`). This means that the parent update to A which happens while B is in progress gets ignored, preventing it from double-firing for B.

I'm a bit averse to updating the default eager condition to contain this logic, as I do think in the general case we _would_ want to listen for / react to parent updates that happen while the asset is in progress (i.e. this is a rather broad brush we're painting with). In a sense, it's trading an edge case for an edge case.

Curious on your thoughts towards this approach.

**Comment by Zan-L** [comment]
I can see it working but it's a weird solution:
1. In_progress() will overshadow newly_updated() because an asset can only be updated during in_progress() (let's ignore runless materialization events for now)
2. Using a status in since is against its principle that you should use events only. 
3. Related to 2, since it's a status, you keep resetting the condition before since() during the asset progress. Unless you have some mechnism to ignore it and reset it only once, it could add many meaningless ticks to show in the automation history. 

What's your take?

**Comment by OwenKephart** [comment]
High level, agree that it's a weird solution, but more granularly:

> In_progress() will overshadow newly_updated() because an asset can only be updated during in_progress()

Definitely correct, that's an explicit drawback of the solution (parent updates while the asset is in progress will be ignored)

> Using a status in since is against its principle that you should use events only

Good point, and I agree that in principle it's odd to have a "status" being used in that capacity. You could avoid this by detecting the "trailing edge" of the in progress condition, e.g. replace `AutomationCondition.in_progress()` with `(~AutomationCondition.in_progress()).newly_true()`, which would have the same behavior. I avoided going that route because it felt unnecessarily complicated given the behavior is the same, but you're correct that that's "more correct".

>  Unless you have some mechanism to ignore it and reset it only once...

The logic that determines if we store / show an evaluation is actually based off of a hash of the properties of the evaluation. So if each individual sub-condition returns the same value, no new tick will be stored. So in this case, the fact that in_progress remains true and continually "resets" the since condition doesn't actually make a difference in terms of how many evaluations exist.

**Comment by Zan-L** [comment]
Really appreciate your feedback - that confirmed that my understanding towards your system is mostly correct. Unfortunately, I already tested with both ((~in_progress()).newly_true()), which I gave in my initial post, and (newly_requested() | (~in_progress()).newly_true()), and neither worked as expected. I'm leaning towards having a new event condition for this case. But feel free to close this issue if you feel it's too much of an edge case.

**Comment by OwenKephart** [comment]
@Zan-L To clarify, did:

```python
... (
    dg.AutomationCondition.newly_missing() | dg.AutomationCondition.any_deps_updated()
).since(
    dg.AutomationCondition.newly_requested()
    | dg.AutomationCondition.newly_updated()
    | (~dg.AutomationCondition.in_progress()).newly_true()
) ...
```

not work for you? (as opposed to simply `.since((~dg.AutomationCondition.in_progress()).newly_true())`)

Another thing to note here is that we're working on a feature here that lets you write your own custom python code as an automation condition, which might be a worthwhile exploration here, although I don't have particular recommendations for what code to write in that case.

But if you want to try that out, all you have to do to enable it is:

```python
defs = Definitions(
    ...,
    sensors=[
        AutomationConditionSensorDefinition(
            "some_name",
            asset_selection="*",
            # this causes the sensor to evaluate on your user code server instead of the daemon
            user_code=True,
        ),
    ]
)
```

From there, you can create subclasses of `AutomationCondition` (requires a couple of private imports for AutomationResult and AutomationContext at the moment though, as this feature hasn't been "officially" released yet).

**Comment by Zan-L** [comment]
In response to your question - No. If you consider the following four combinations:
1. ((~in_progress()).newly_true())
2. (newly_requested() | (~in_progress()).newly_true())
3. (newly_updated() | (~in_progress()).newly_true())
4. (newly_requested() | newly_updated() | (~in_progress()).newly_true())

Only 1 provided the relatively most reliable outcome. Even so, it still can't guarantee anything especially for long-running assets (> 15 minutes). In addition, it's simply not appropriate to use as there will be infinite loops for failed ones. 

In short, whether ((~in_progress()).newly_true()) gets evaluated correctly seems to be interfered with both the timing of asset functions and other conditions. 

I speculate that the latter is caused by state hasing. If newly_requested() is present, it will reset the condition and cause it to evaluate to true during the short time when the upstream is updated but itself is still in-progress. Removing newly_requested() indirectly improves it, possibly due to not resetting the condition and thus postponing the evaluation until when it finishes.

**Comment by Zan-L** [comment]
@OwenKephart Just found out that this relates to another issue that's worth fixing - in the case where an initial run (let's say again with assets A, B, and C) failed (none was updated), if we re-execute the run, then some runs of (B, C) could be misfired because the failed run does not have any data update, just like when you have an optional output.

**Comment by OwenKephart** [comment]
Hi @Zan-L , thanks for your patience on this. Do you have an example that can (somewhat) reliably reproduce the issue that you're seeing? With my local setup, I have not been able to observe examples of the behavior you're describing, but I suspect it's due to specific timing choices I'm making in my tests.

I think you're correct that there's a timing issue at play here, and I've noticed an issue in how we calculate the in progress status for unpartitioned assets which may solve the inconsistency issue, but it's hard to confirm if it works or not without a repro available.

I think conceptually the solution of ignoring parent updates that happen as an asset changes state (in_progress -> not in progress or not failed -> failed) feels logically correct -- can you elaborate a bit more on what difference this would make in the case of long-running assets vs. short ones?

**Comment by OwenKephart** [comment]
Ah another way of approaching this that I think is simpler overall is to add a new condition that detects if the latest run of an asset included the target child asset. So `any_deps_match((newly_updated() & ~run_contains_child())`. This way, with A -> B -> C, If we launch a run A -> B -> C, and A updates within that run, B will ignore this update.

I'll explore that direction as well.

**Comment by Zan-L** [comment]
I saw you already had a fix for it. Nice work! Looking forward to testing it. 
Sorry I couldn't get back to you with more reproducible examples.

Best,
ZL

**Comment by Zan-L** [comment]
> I think conceptually the solution of ignoring parent updates that happen as an asset changes state (in_progress -> not in progress or not failed -> failed) feels logically correct

@OwenKephart I agree. And now I think the reason why it did not work may be that **in_progress() (or how it is evaluated) is bugged**. I just noticed today that for a pair of assets with outputs, (A->B), after a combined run of (A, B) is spawned, a redundant run of just (B) can still be fired while (A, B) is still running. 

Here are 5 runs for my B:
![image](https://github.com/user-attachments/assets/0876fdd1-28e8-4da2-93fb-72934a7055af)

But there should only be 3 because there were only 3 (A, B) runs. If you look closely, the first and the fourth materialization was mis-fired.

Here is what the automation sensor ticks for B should look like at 7:44 pm:
![image](https://github.com/user-attachments/assets/31b5fa6d-a2e5-4ad3-ad2b-543b6cdd3285)

Note that the `Not in_progress()` should produce 0 partition. However, for 4:40 am and 6:40 am, the same fourth tick went wrong:
![image](https://github.com/user-attachments/assets/609f86da-61df-4bf1-a9e4-efbdaf804130)

Again, in this case, both A and B return outputs, so I'm sure it's caused by the in_progress() condition.
